File: Assets/Game/Scripts/Moduls/P_Extra_Point.lua
local GameSoundType = Ballance2.Services.GameSoundType
local ObjectStateBackupUtils = Ballance2.Utils.ObjectStateBackupUtils
local SmoothFly = Ballance2.Game.Utils.SmoothFly
local SmoothFlyType = Ballance2.Game.Utils.SmoothFlyType
local Physics = UnityEngine.Physics
local Vector3 = UnityEngine.Vector3
local AudioSource = UnityEngine.AudioSource

---@class P_Extra_Point : ModulBase 
---@field P_Extra_Point_Tigger TiggerTester
---@field P_Extra_Point_Floor GameObject
---@field P_Extra_Point_Ball0 GameObject
---@field P_Extra_Point_Ball1 GameObject
---@field P_Extra_Point_Ball2 GameObject
---@field P_Extra_Point_Ball3 GameObject
---@field P_Extra_Point_Ball4 GameObject
---@field P_Extra_Point_Ball5 GameObject
---@field P_Extra_Point_Ball6 GameObject
---@field P_Extra_Point_Ball_Povit1 GameObject
---@field P_Extra_Point_Ball_Povit2 GameObject
---@field P_Extra_Point_Ball_Povit4 GameObject
---@field P_Extra_Point_Ball_Povit5 GameObject
---@field P_Extra_Point_Ball_Povit6 GameObject
---@field P_Extra_Point_Fizz GameObject
---@field P_Extra_Point_Sound AudioSource
P_Extra_Point = ModulBase:extend()

function P_Extra_Point:new()
  P_Extra_Point.super.new(self)
  self._Actived = false
  self._RotDegree = 6
  self._Rotate = false
  self._FlyUpTime = 2.5
  self._FlyFollowTime = 0.3
  self.AutoActiveBaseGameObject = false
  self.EnableBallRangeChecker = true
  self.BallCheckeRange = 80
end

function P_Extra_Point:Start()
  ModulBase.Start(self)
  
  if not self.IsPreviewMode then
    --初始化xiao'q
    self._P_Extra_Point_Ball_Fly1 = self.P_Extra_Point_Ball1:GetComponent(SmoothFly) ---@type SmoothFly
    self._P_Extra_Point_Ball_Fly2 = self.P_Extra_Point_Ball2:GetComponent(SmoothFly) ---@type SmoothFly
    self._P_Extra_Point_Ball_Fly3 = self.P_Extra_Point_Ball3:GetComponent(SmoothFly) ---@type SmoothFly
    self._P_Extra_Point_Ball_Fly4 = self.P_Extra_Point_Ball4:GetComponent(SmoothFly) ---@type SmoothFly
    self._P_Extra_Point_Ball_Fly5 = self.P_Extra_Point_Ball5:GetComponent(SmoothFly) ---@type SmoothFly
    self._P_Extra_Point_Ball_Fly6 = self.P_Extra_Point_Ball6:GetComponent(SmoothFly) ---@type SmoothFly

    for i = 1, 6, 1 do
      local fly = self['P_Extra_Point_Ball'..i]:GetComponent(SmoothFly) ---@type SmoothFly
      local hitAudio = self.transform:Find('P_Extra_Point_Ball'..i..'/P_Extra_Point_Hit'):GetComponent(AudioSource) ---@type AudioSource
      local hitFizz = self.transform:Find('P_Extra_Point_Ball'..i..'/P_Extra_Point_Fizz').gameObject
      local ballParticle = self.transform:Find('P_Extra_Point_Ball'..i..'/P_Extra_Point_Ball').gameObject
      local flowParticle = self.transform:Find('P_Extra_Point_Ball'..i..'/P_Extra_Point_Flow').gameObject

      Game.SoundManager:RegisterSoundPlayer(GameSoundType.Normal, hitAudio)
      self['_P_Extra_Point_Ball_Fly'..i] = fly
      self['_P_Extra_Point_Ball_Rest'..i] = function ()
        ballParticle:SetActive(true)
        flowParticle:SetActive(true)
        hitFizz:SetActive(false)
        hitAudio:Stop()
      end

      fly.StopWhenArrival = true
      fly.ArrivalDiatance = 2
      fly.ArrivalCallback = function ()
        if not self._FlyUp then
          ballParticle:SetActive(false)
          flowParticle:SetActive(false)
          hitFizz:SetActive(true)
          hitAudio:Play()
          GamePlay.GamePlayManager:AddPoint(20) --小球是20分
        end
      end
    end

    ---@param otherBody GameObject
    self.P_Extra_Point_Tigger.onTriggerEnter = function (_, otherBody)
      if not self._Actived and otherBody.tag == 'Ball' then
        self._Actived = true
        self:StartFly()
        GamePlay.GamePlayManager:AddPoint(100) --大球是100分
      end
    end
  end

  self._OnFloor = false
  --触发射线，检查当前下方是不是路面，如果是，则显示 Shadow 
  ---@type boolean
  local ok, 
  ---@type RaycastHit
  hitinfo = Physics.Raycast(self.transform.position, Vector3(0, -1, 0), Slua.out, 5) 
  if ok and hitinfo.collider ~= nil then
    local parentName = hitinfo.collider.gameObject.tag
    if (parentName == 'Phys_Floors' or parentName == 'Phys_FloorWoods') and  parentName ~= 'Phys_FloorRails' then
      self._OnFloor = true
    end
  end

  self._RotCenter = self.transform.position
  self._RotAxis1 = self.P_Extra_Point_Ball_Povit1.transform.up
  self._RotAxis2 = self.P_Extra_Point_Ball_Povit2.transform.up
  self._RotAxis4 = self.P_Extra_Point_Ball_Povit4.transform.up
  self._RotAxis5 = self.P_Extra_Point_Ball_Povit5.transform.up
  self._RotAxis6 = self.P_Extra_Point_Ball_Povit6.transform.up
  self.P_Extra_Point_Ball_Povit1:SetActive(false)
  self.P_Extra_Point_Ball_Povit2:SetActive(false)
  self.P_Extra_Point_Ball_Povit4:SetActive(false)
  self.P_Extra_Point_Ball_Povit5:SetActive(false)
  self.P_Extra_Point_Ball_Povit6:SetActive(false)
end
function P_Extra_Point:StartFly()
  self._Rotate = false
  self._FlyUp = true

  self.P_Extra_Point_Floor:SetActive(false)
  self.P_Extra_Point_Ball0:SetActive(false)
  self.P_Extra_Point_Fizz:SetActive(true)
  self.P_Extra_Point_Sound:Play()

  local upY = 16

  for i = 1, 6, 1 do
    local ball = self['P_Extra_Point_Ball'..i] ---@type GameObject
    local fly = self['_P_Extra_Point_Ball_Fly'..i] ---@type SmoothFly
    local posMult = Vector3(1.2, 1, 1.2)
    local localPos = Vector3.Scale(ball.transform.localPosition, posMult)
  
    fly.Type = SmoothFlyType.Lerp
    fly.TargetTransform = nil
    fly.TargetPos = self.transform:TransformPoint(localPos.x, upY, localPos.z)
    fly.Time = self._FlyUpTime
    fly.Fly = true
    upY = upY + 0.2
  end

  local fTime =  self._FlyFollowTime

  LuaTimer.Add(1250, function ()
    self.P_Extra_Point_Fizz:SetActive(false)
    self._FlyUp = false

    local followTarget = GamePlay.CamManager.Target
    for i = 1, 6, 1 do
      local fly = self['_P_Extra_Point_Ball_Fly'..i] ---@type SmoothFly
      fly.Fly = true
      fly.Type = SmoothFlyType.SmoothDamp
      fly.TargetTransform = followTarget
      fly.Time = fTime
      fTime = fTime - 0.02
    end
  end)
end
function P_Extra_Point:Update()
  --旋转小球
  if self._Rotate then
    self.P_Extra_Point_Ball1.transform:RotateAround(self._RotCenter, self._RotAxis1, self._RotDegree)
    self.P_Extra_Point_Ball2.transform:RotateAround(self._RotCenter, self._RotAxis2, self._RotDegree)
    self.P_Extra_Point_Ball3.transform:RotateAround(self._RotCenter, self._RotAxis1, -self._RotDegree)
    self.P_Extra_Point_Ball4.transform:RotateAround(self._RotCenter, self._RotAxis4, self._RotDegree)
    self.P_Extra_Point_Ball5.transform:RotateAround(self._RotCenter, self._RotAxis5, self._RotDegree)
    self.P_Extra_Point_Ball6.transform:RotateAround(self._RotCenter, self._RotAxis6, self._RotDegree)
  end 
end

function P_Extra_Point:Active()
  ModulBase.Active(self)
  if not self._Actived then
    self.gameObject:SetActive(true)
    self.P_Extra_Point_Floor:SetActive(self._OnFloor)
    self.P_Extra_Point_Ball0:SetActive(true)
    self.P_Extra_Point_Ball1:SetActive(true)
    self.P_Extra_Point_Ball2:SetActive(true)
    self.P_Extra_Point_Ball3:SetActive(true)
    self.P_Extra_Point_Ball4:SetActive(true)
    self.P_Extra_Point_Ball5:SetActive(true)
    self.P_Extra_Point_Ball6:SetActive(true)
    self._FlyModUp = false
    self._FlyModFollow = false
  end
end
function P_Extra_Point:Deactive()
  ModulBase.Deactive(self)
  self._Rotate = false
  self.gameObject:SetActive(false)
end
function P_Extra_Point:Reset(type)
  if(type == 'levelRestart') then
    self._Actived = false
    ObjectStateBackupUtils.RestoreObject(self.P_Extra_Point_Ball1)
    ObjectStateBackupUtils.RestoreObject(self.P_Extra_Point_Ball2)
    ObjectStateBackupUtils.RestoreObject(self.P_Extra_Point_Ball3)
    ObjectStateBackupUtils.RestoreObject(self.P_Extra_Point_Ball4)
    ObjectStateBackupUtils.RestoreObject(self.P_Extra_Point_Ball5)
    ObjectStateBackupUtils.RestoreObject(self.P_Extra_Point_Ball6)
    self._P_Extra_Point_Ball_Fly1.Fly = false
    self._P_Extra_Point_Ball_Fly2.Fly = false
    self._P_Extra_Point_Ball_Fly3.Fly = false
    self._P_Extra_Point_Ball_Fly4.Fly = false
    self._P_Extra_Point_Ball_Fly5.Fly = false
    self._P_Extra_Point_Ball_Fly6.Fly = false
    for i = 1, 6, 1 do
      self['_P_Extra_Point_Ball_Rest'..i]()
    end
  end
end
function P_Extra_Point:Backup()
  ObjectStateBackupUtils.BackUpObject(self.P_Extra_Point_Ball1)
  ObjectStateBackupUtils.BackUpObject(self.P_Extra_Point_Ball2)
  ObjectStateBackupUtils.BackUpObject(self.P_Extra_Point_Ball3)
  ObjectStateBackupUtils.BackUpObject(self.P_Extra_Point_Ball4)
  ObjectStateBackupUtils.BackUpObject(self.P_Extra_Point_Ball5)
  ObjectStateBackupUtils.BackUpObject(self.P_Extra_Point_Ball6)
end

function P_Extra_Point:ActiveForPreview()
  self:Active()
end
function P_Extra_Point:DeactiveForPreview()
  self:Deactive()
end

function P_Extra_Point:BallEnterRange()
  if not self._Actived and not self._Rotate then --进入范围后才旋转
    self._Rotate = true
  end
end
function P_Extra_Point:BallLeaveRange()
  if not self._Actived and self._Rotate then --离开范围后停止旋转
    self._Rotate = false
  end
end

function CreateClass:P_Extra_Point()
  return P_Extra_Point()
end

File: Assets/Game/GamePlay/Scripts/Moduls/P_Extra_Point.cs
using Ballance2.Game.Utils;
using Ballance2.Services;
using Ballance2.Utils;
using UnityEngine;

namespace Ballance2.Game.GamePlay.Moduls
{
  /// <summary>
  /// 额外分数组件
  /// </summary>
  public class P_Extra_Point : ModulBase 
  {
    public TiggerTester P_Extra_Point_Tigger;
    public GameObject P_Extra_Point_Floor;
    public GameObject[] P_Extra_Point_Ball = new GameObject[7];
    public SmoothFly[] P_Extra_Point_Target = new SmoothFly[7];
    public GameObject P_Extra_Point_Ball_Povit1;
    public GameObject P_Extra_Point_Ball_Povit2;
    public GameObject P_Extra_Point_Ball_Povit4;
    public GameObject P_Extra_Point_Ball_Povit5;
    public GameObject P_Extra_Point_Ball_Povit6;
    public GameObject P_Extra_Point_Fizz;
    public AudioSource P_Extra_Point_Sound;

    public bool _Rotate = false;
    public bool _FlyUp = false;
    public bool _FlyFollow = false;
    public float _RotDegree = 6f;
    public float _FlyUpTime = 0.4f;
    public float _FlyFollowTime = 0.3f;
    public float _FlyTargetUpTime = 0.3f;
    public float _FlyTargetFollowTime = 0.2f;

    private SmoothFly[] _P_Extra_Point_Ball_Fly = new SmoothFly[7];
    private GameManager.VoidDelegate[] _P_Extra_Point_Ball_Rest = new GameManager.VoidDelegate[7];
    private Vector3 _RotCenter;
    private Vector3 _RotAxis1;
    private Vector3 _RotAxis2;
    private Vector3 _RotAxis4;
    private Vector3 _RotAxis5;
    private Vector3 _RotAxis6;

    private bool _Actived = false;
    private bool _OnFloor = false;

    public P_Extra_Point() {
      AutoActiveBaseGameObject = false;
      EnableBallRangeChecker = true;
      BallCheckeRange = 200;
    }

    protected override void Start()
    {
      if (!IsPreviewMode) {
        //初始化小球
        for (int i = 1; i <= 6; i++)
          _P_Extra_Point_Ball_Fly[i] = P_Extra_Point_Ball[i].GetComponent<SmoothFly>();

        for(int i = 1; i <= 6; i++) {
          var fly = _P_Extra_Point_Ball_Fly[i];
          var hitAudio = transform.Find($"P_Extra_Point_Ball{i}/P_Extra_Point_Hit").GetComponent<AudioSource>();
          var hitFizz = transform.Find($"P_Extra_Point_Ball{i}/P_Extra_Point_Fizz").gameObject;
          var ballParticle = transform.Find($"P_Extra_Point_Ball{i}/P_Extra_Point_Ball").gameObject;
          var flowParticle = transform.Find($"P_Extra_Point_Ball{i}/P_Extra_Point_Flow").gameObject;
          var target = P_Extra_Point_Target[i];

          GameSoundManager.Instance.RegisterSoundPlayer(GameSoundType.Normal, hitAudio);
          _P_Extra_Point_Ball_Rest[i] = () => {
            ballParticle.SetActive(true);
            flowParticle.SetActive(true);
            hitFizz.SetActive(false);
            hitAudio.Stop();
          };
          
          fly.Type = SmoothFlyType.SmoothDamp;
          fly.TargetTransform = target.transform;
          fly.StopWhenArrival = true;
          fly.ArrivalDiatance = 2;
          fly.ArrivalCallback = (fly) => {
            if (!_FlyUp) {
              target.Fly = false; //停止追逐
              ballParticle.SetActive(false);
              flowParticle.SetActive(false);
              hitFizz.SetActive(true);
              hitAudio.Play();
              GamePlayManager.Instance.AddPoint(20); //小球是20分
            }
          };
        }

        P_Extra_Point_Tigger.onTriggerEnter = (_, otherBody) => {
          if (IsActive && ! _Actived && otherBody.tag == "Ball") {
            _Actived = true;
            StartFly();
            GamePlayManager.Instance.AddPoint(100); //大球是100分
          }
        };
      }
      //触发射线，检查当前下方是不是路面，如果是，则显示 Shadow
      RaycastHit hitinfo;
      if (Physics.Raycast(transform.position, new Vector3(0, -1, 0), out hitinfo, 5) && hitinfo.collider != null) {
        var parentName = hitinfo.collider.gameObject.tag;
        if (parentName == "Phys_Floors" || (parentName == "Phys_FloorWoods" && parentName != "Phys_FloorRails"))
          _OnFloor = true;
        else
          _OnFloor = false;
      }
      else _OnFloor = false;

      _RotCenter = transform.position;
      _RotAxis1 = P_Extra_Point_Ball_Povit1.transform.up;
      _RotAxis2 = P_Extra_Point_Ball_Povit2.transform.up;
      _RotAxis4 = P_Extra_Point_Ball_Povit4.transform.up;
      _RotAxis5 = P_Extra_Point_Ball_Povit5.transform.up;
      _RotAxis6 = P_Extra_Point_Ball_Povit6.transform.up;
      P_Extra_Point_Ball_Povit1.SetActive(false);
      P_Extra_Point_Ball_Povit2.SetActive(false);
      P_Extra_Point_Ball_Povit4.SetActive(false);
      P_Extra_Point_Ball_Povit5.SetActive(false);
      P_Extra_Point_Ball_Povit6.SetActive(false);

      base.Start();
    }

    private void StartFly() {
      _Rotate = false;
      _FlyUp = true;

      P_Extra_Point_Floor.SetActive(false);
      P_Extra_Point_Ball[0].SetActive(false);
      P_Extra_Point_Fizz.SetActive(true);
      P_Extra_Point_Sound.Play();

      //目标平移到指定位置

      var upY = 30.0f;

      for(int i = 1; i <= 6; i++) {

        var ball = P_Extra_Point_Ball[i];
        var flyTarget = P_Extra_Point_Target[i];
        var fly = _P_Extra_Point_Ball_Fly[i];
        var posMult = new Vector3(1.2f, 1.0f, 1.2f);
        var pos = ball.transform.localPosition;
        var localPos = Vector3.Scale(pos, posMult);

        flyTarget.transform.localPosition = pos;
        flyTarget.CurrentVelocity = Vector3.zero;
        flyTarget.TargetTransform = null;
        flyTarget.TargetPos = transform.TransformPoint(localPos.x, upY, localPos.z);
        flyTarget.Fly = true;
        flyTarget.Time = _FlyTargetUpTime;

        fly.Time = _FlyUpTime;
        fly.CurrentVelocity = Vector3.zero;
        fly.StopWhenArrival = false;
        fly.Fly = true;

        upY = upY + 0.3f;
      }

      var fTime =  _FlyFollowTime;

      GameTimer.Delay(0.750f, () => {
        for(int i = 1; i <= 6; i++) {
          var fly = P_Extra_Point_Target[i];
          fly.TargetPos = Vector3.zero;
          fly.TargetTransform = GamePlayManager.Instance.CamManager.Target;
          fly.Time = _FlyTargetFollowTime;
          fly.CurrentVelocity = Vector3.zero;
        }
      });
      GameTimer.Delay(1, () => {
        P_Extra_Point_Fizz.SetActive(false);
        _FlyUp = false;
        for(int i = 1; i <= 6; i++) {
          var fly = _P_Extra_Point_Ball_Fly[i];
          fly.StopWhenArrival = true;
          fly.Fly = true;
          fly.Time = fTime;
          fTime = fTime - 0.03f;
        }
      });
    }

    private void Update() {
      //旋转小球
      if (_Rotate) {
        var rotate = _RotDegree * Time.deltaTime;
        P_Extra_Point_Ball[1].transform.RotateAround(_RotCenter, _RotAxis1, rotate);
        P_Extra_Point_Ball[2].transform.RotateAround(_RotCenter, _RotAxis2, rotate);;
        P_Extra_Point_Ball[3].transform.RotateAround(_RotCenter, _RotAxis1, -rotate);
        P_Extra_Point_Ball[4].transform.RotateAround(_RotCenter, _RotAxis4, rotate);
        P_Extra_Point_Ball[5].transform.RotateAround(_RotCenter, _RotAxis5, rotate);
        P_Extra_Point_Ball[6].transform.RotateAround(_RotCenter, _RotAxis6, rotate);
      } 
    }

    public override void Active()
    {
      base.Active();

      if (!_Actived) {
        gameObject.SetActive(true);
        P_Extra_Point_Floor.SetActive(_OnFloor);
        for(int i = 1; i <= 6; i++)
          P_Extra_Point_Ball[i].SetActive(true);
        _FlyUp = false;
        _FlyFollow = false;
        if (BallInRange)
          _Rotate = true;
      }
    }
    public override void Deactive()
    {
      base.Deactive();

      _Rotate = false;
      gameObject.SetActive(false);
    }
    public override void Reset(ModulBaseResetType type)
    {
      if (type == ModulBaseResetType.LevelRestart) {
        _Actived = false;
        P_Extra_Point_Ball[0].SetActive(true);
        for (int i = 1; i <= 6; i++) {
          _P_Extra_Point_Ball_Fly[i].Fly = false;
          ObjectStateBackupUtils.RestoreObject(P_Extra_Point_Ball[i]);
          _P_Extra_Point_Ball_Rest[i]();
        }
      }
    }
    public override void Backup()
    {
      for (int i = 1; i <= 6; i++) 
        ObjectStateBackupUtils.BackUpObject(P_Extra_Point_Ball[i]);
    }
    public override void ActiveForPreview()
    {
      this.Active();
    }
    public override void DeactiveForPreview()
    {
      this.Deactive();
    }
    protected override void BallEnterRange() {
      if (IsActive && ! _Actived && !_Rotate) //进入范围后才旋转
        _Rotate = true;
    }   
    protected override void BallLeaveRange() {
      if (IsActive && !_Actived && _Rotate) //离开范围后停止旋转
        _Rotate = false;
    }
  }
}